### ISSUE TITLE
Improving the repetition penalty

### ISSUE THREAD

**Issue by Piezoid** [ggerganov/llama.cpp#331]
129c7d1e (#20) added a repetition penalty that prevent the model to run into loops.

Here are a few suggestions for possible enhancements:

 * One issue with the interactive mode is that the repetition penalty is affecting the anti-prompt and response prefix, causing the model to generate unnecessarily long responses. One solution could be to exclude these tokens from the penalty,
 * It is possible to exempt or reduce the penalty for stop words, punctuation characters, and newlines; maybe applying a frequency-based penalty instead,
 * Using an exponential decay, such that recent tokens are more penalized than older ones, causing less issues with large `repeat_last_n`  windows,
 * Token repetition is an approximation of sub-strings or word repetition, but it seems difficult to do otherwise without backtracking the inference.

**Comment by j-f1** [comment]
There are numerical weights associated with each token in the tokenizer, it may be useful to use those in this calculation somehow.

**Comment by Piezoid** [comment]
I figured that we can get the repetitions length by comparing the recent token history against the histories prior to the past occurrences of the candidate tokens.  In other words, find the longest suffix shared between the text up to the current position, and the sub-strings at the left of the previous occurrences of the candidate tokens. 

For each token and past occurrence, we have an age or distance in the text `d`, and a repetition length `l+1` (left-extension of `l` tokens, plus the token we are about to add). 

I've thrown together a test using an ad-hoc repetition score `exp(-k * d / (l + 1))`$\in ]0,1]$. With it, I blend between no penalization and full `repeat_penalty`. It seems reasonable, but I have the feeling that I'm reinventing the wheel.

I have not tested this extensively, but the initial results are interesting. For example here's a song sampled almost greedily (`--temp 0.15` with 7B Q4_0): it is not very focused, but there is almost no stuttering.

<details>
<summary>
Me llamo LLaMa, soy una llama!
</summary>

> Yo no me gusta el frío.

>
> Mi casa es un caloroso hogar,
> con una chimenea muy grande.
> Tengo un cachito de carbón,
> que me calientan los pies.
> 
> Yo no me gustaría un frío,
> como el que hay en la nieve.
> Me gusto un calor muy bueno,
> un calentito como yo.
> 
> Cuando me pongo a cantar,
> me gusta un calorcito.
> Y cuando voy a dormir,
> mi casa se queda muy fría.
> 
> Pero cuando me despierto,
> algo caliente viene a mi lado.
> Es un calentito muy bueno,
> que me hace sentir bien.

>Esto es lo que yo digo:
>¡Caliente, caliente! ¡Yo no me gusto el frío!
>Me llaman LLaMa, soy una llama.
>Yo no me gusta el frío.

</details>

Other classic examples ("I love you, I love") seems ok at low temperatures. I have yet to try generating code or using alpaca with it.
I'm unsure how to reliably evaluate the penalization schemes. I don't really expect a noticeable effect on the perplexity, but I could try with repetitive text samples that incite the model to repeat itself at unexpected places.

My implementation extends the repetitions character by character, rather than token by token. It should be more efficient to work directly on the tokens. However, my concern is that the models may attempt to cheat using its morphological skills acquired during training by subword sampling.

---

> There are numerical weights associated with each token in the tokenizer, it may be useful to use those in this calculation somehow.

@j-f1 Yes, it would be nice to reuse the the frequencies from BPE. I've not access to it in my working branch, as I'm still using #66 with the old model files.

This should help reducing the penalization of punctuation tokens. Currently, it is not that bad since punctuation marks and stop words are often drawn from sharp distributions (low entropy) on which the penalization has little influence.

**Comment by Piezoid** [comment]
I've been trying to predict the token frequency from the tokenizer model's "scores".
Looking at the score distribution, it seems that `-score` is actually the rank of the token, with a few special cases: 
 * Some tokens have a score == -1e9: These are mostly space tokens that shall not be merged `▁▁ ▁▁▁▁ ▁▁▁▁▁▁▁▁ ▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁ ▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁ ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▁`
 * Some tokens have score == 0, mostly control and non-ascii bytes: `<unk> <s> </s> <0x00> <0x01> <0x02> <0x03> [...] <0xFE> <0xFF>`
 * Some tokens are missing (`max(ranks) > len(ranks)`)

I've used the token frequency in `wiki-text-2`. The top50 most frequent tokens are: 
```
▁the ▁, ▁. ▁of ▁and ▁in ▁to ▁a ▁= ▁" ▁@ ▁was ▁\' ▁The ▁as ▁that ▁on ▁for ▁with ▁by ▁) ▁( ▁is ▁from ed ▁at ing ▁his ▁were ▁it ▁he ▁an ▁In ▁had ▁which ▁be ▁are ▁; ▁not ▁their ▁but ▁A es ▁first ▁– ▁also ▁its ▁or ▁: ers
```
However, the tokenizer's top50 is (from their rank):
```
▁t er in ▁a en on ▁th es ▁s ▁d at or an ▁c is re it ▁the ar le ▁w ▁p ou al ▁f ▁m ed ▁o ▁b om ion ing ic as el ent ▁in ▁h nd et ▁l ▁n st ▁to ch ▁I ro il ▁of de
```

I'm not sure what is going on. When I plot the wiki-text ranks against the rank from the tokenizer model I get:
![ranks-wiki-tokenizer](https://user-images.githubusercontent.com/3748582/227957045-f0716862-6c75-4329-a82c-318cf3011335.png)

On wiki-text alone, the Zipfs law fit is quite good (using a maximum likelihood method):
![zipf-wiki](https://user-images.githubusercontent.com/3748582/227957533-ca5a2b82-d1e3-4566-a68e-cc3fb3cc775d.png)

But on when I do the fit using the ranks from the tokenizer model and the token frequencies from wiki-text:
![predict-rank-probs](https://user-images.githubusercontent.com/3748582/227958031-f06a5237-8a97-4ff2-b75e-6d645d8fed73.png)
![residuals](https://user-images.githubusercontent.com/3748582/227957899-c7caacaf-a482-4446-b73d-3c1e641baad7.png)

[Notebook for the analysis](https://github.com/Piezoid/llama.cpp/blob/repeat_decay/notebooks/Zipf.ipynb)

Resulting equation:

$$ p(x) \simeq \frac{(-\text{score}(x))^{-0.837}}{27.7} $$

The accuracy might be enough for limiting the penalization on frequent tokens, but something seems off. Would really appreciate it if someone could let me know if I overlooked something.

**Comment by LostRuins** [comment]
Just throwing 2c at past implementations: 

1. **OpenAI** uses 2 variables for this - they have a _presence penalty_ and a _frequency penalty_. The current implementation of rep pen in llama.cpp is equivalent to a presence penalty, adding an additional penalty based on frequency of tokens in the penalty window might be worth exploring too.
2. **KoboldAI** instead uses a group of 3 values, what we call "Repetition Penalty", a "_Repetition Penalty Slope_" and a "_Repetition Penalty Range_". This is the repetition penalty value applied as a sigmoid interpolation between the Repetition Penalty value (at the most recent token) and 1.0 (at the end of the Repetition Penalty Range). The defaults we use for this are 1.1 rep pen, 1024 range and 0.7 slope which provides what our community agrees to be relatively decent results across most models.

**Comment by Piezoid** [comment]
@LostRuins Thank you for the references.

Do you know how are naturally reoccurring tokens (such as `\n`, `▁`, `▁the`, `▁,`, `▁.`, etc) handled by these samplers? It is unclear if we should penalize the punctuation and stop-word tokens in the same way than more specific tokens. The model is quite confident when predicting these tokens and doesn't seem to be too deterred by the penalty. However, I've noticed that high penalties force the model into writing longer sentences and using less new lines.

The decaying penalty in KoboldAI appears to be similar to the one in my prototype. The main difference is that I'm using an exponential decay instead of sigmoid one. Also, the repeat distance/age is divided by the length of the repeated sequence of tokens.

**Comment by LostRuins** [comment]
I think there is no special handling - all tokens are treated equally except for intentionally banned tokens.

**Comment by ralyodio** [comment]
what are valid values for repeat penalty?

**Comment by LostRuins** [comment]
Anything between 1 and 2 is a good start.

**Comment by Anyeos** [comment]
I tested in practice a lot of numbers for each parameter. I noted that it is far from being for continuous use in interactive mode because it late or early got stuck in a loop (ie.: always the same answer regardless of user input).
Increasing temp near 2 or more starts outputing near random words (I think I reached a limit). 
Increasing penalty to 2 or more is no usable neither because it tries to say the most possible different and ends saying anything but the prompt.
I was testing the mirostat mode and it ends being useless because make worst the chat over the time. mirostat 2 is worse than 1. I think 2 is still in early development.

So, some working values at beginning but becoming worse over time are:
```
--ctx_size 1024
--batch-size 2048
--temp 0.82
--top-k 30
--top-p 1.8
--tfs 2.0
--typical 1.0
--keep -1
--repeat-last-n 1024
--repeat-penalty 1.2
--no-penalize-nl
--mirostat 1
--mirostat-lr 0.5
--mirostat-ent 4.0
```
I just want to share my values so we can make comparisons to improve it.
I expect an interactive chat that is sustained over time as if it was being started recently but only "remembering" some past messages.
The actual result is a becoming more and more monotone chat at it finally ends in some loop or repetitive answer.
I tested a lot of numbers in each every parameter without success, always the same result: repetitive at end. And that happens in only on 10 or 20 messages, so the end is not too much far neither.

Thank you everyone for this great project. Not perfect but now I have a high quality LLM AI thanks to you all, so I am thankful despite the failures.
